common/ulsr_hil_client.py

```diff
@@ -94,35 +92,42 @@ def hil_reserve_nodes(nodelist, from_project, hil_client=None):
 
         # if node already in the free pool, skip any processing.
         if project is None:
-            log_info('HIL release: Node `%s` already in the free pool, skipping' % node)
+            log_info(
+                'HIL release: Node `%s` already in the free pool, skipping' % node)
             nodelist.remove(node)
-
         elif project != from_project and project != MAINTENANCE_PROJECT:
-            log_error('HIL reservation failure: Node `%s` (in project `%s`) not in `%s` project' % (node, project, from_project))
+            log_error('HIL reservation failure: Node `%s` (in project `%s`) not in `%s` project' % (
+                node, project, from_project))
             raise ProjectMismatchError()
 
     # Power off all nodes.
     # Check if the obm network is connected, if not, connect it.
     for node in nodelist:
         if not _is_network_connected(hil_client, node, OBM_NETWORK, 'vlan/native'):
-            hil_client.node.connect_network(node, OBM_NIC, OBM_NETWORK, 'vlan/native')
-        _ensure_network_connected(hil_client, node, OBM_NETWORK, 'vlan/native')
+            status_id = connect_network(hil_client, node, OBM_NIC,
+                                        OBM_NETWORK, 'vlan/native')
+            assert_network_operation_success(hil_client, status_id)
         power_off_node(hil_client, node)
 
-    # Remove all networks from nodes.
-    for node in nodelist:
-        _remove_all_networks(hil_client, node)
-
-    # Finally, remove node from `from_project` and MAINTENANCE-PROJECT.
+    # Queue the network operations to remove all networks from all nodes, but
+    # don't check the success of the operations just yet.
+    status_ids = []
     for node in nodelist:
+        status_ids.extend(remove_all_networks(hil_client, node))
 
-        # this call will check multiple times before raising an error.
-        _ensure_no_networks(hil_client, node)
-        remove_node_from_project(hil_client, node, from_project)
+    # Check the status of the networking actions after we have queued all
+    # operations in the previous step.
+    for status_id in status_ids:
+        assert_network_operation_success(hil_client, status_id)
 
+    # Finally, remove node from `from_project` and MAINTENANCE_PROJECT.
+    for node in nodelist:
         node_info = show_node(hil_client, node)
         project = node_info['project']
 
+        if project == from_project:
+            remove_node_from_project(hil_client, node, from_project)
+
         if project == MAINTENANCE_PROJECT:
             remove_node_from_project(hil_client, node, MAINTENANCE_PROJECT)
```

Comment:
yeah, that sounds better. I'll do that.

Reply:
This is actually right. Now I remember why I did that. Because if the node was in a project, then calling remove will put it in the maintenance project and then you have to call remove again to finally put it in the free pool. And in case the node was already in the maintenance project, it skips the first if clause and goes to the second one. Does that make sense?